runtime/lua/vim/diagnostic.lua

```diff
@@ -1643,6 +1657,42 @@ M.handlers.underline = {
   end,
 }
 
+--- Returns the text to show for the given diagnostics, grouped by columns.
+---
+--- If `opts.as_inlay` is false or nil (default) then all diagnostics are
+--- grouped on the 0-th column, as was previously done.
+---
+--- @param line_diagnostics vim.Diagnostic[]
+--- @param opts vim.diagnostic.Opts.VirtualText
+--- @return table<integer, string[]> chunks_per_column
+local function get_virtual_text_chunks_per_column(line_diagnostics, opts)
+  local group_diags = {} --- @type table<integer, vim.Diagnostic[]>
+
+  if opts.as_inlay then
+    --- Gather diagnostics per-column:
+    for _, diag in ipairs(line_diagnostics) do
+      local col = diag.end_col or diag.col
+      local col_diags = group_diags[col]
+      if col_diags == nil then
+        col_diags = {}
+        group_diags[col] = col_diags
+      end
+      table.insert(col_diags, diag)
```

Comment:
```suggestion
      local col = diag.end_col or diag.col
      group_diags[col] = group_diags[col] or {}
      table.insert(group_diags[col], diag)
```

Reply:
Ah, I was following the style of `diagnostic_lines()` [as seen here](https://github.com/neovim/neovim/blob/487112d6749d4b7a2b82a60fb30cfb9e21c78394/runtime/lua/vim/diagnostic.lua#L654).

Will change it and push once all tests are finished.